components/style/properties/helpers/animated_properties.mako.rs

```diff
@@ -757,6 +766,695 @@ impl Interpolate for LengthOrNone {
+                perspective_matrix = ComputedMatrix {
+                    m11: perspective_matrix.m11, m12: perspective_matrix.m21,
+                    m13: perspective_matrix.m31, m14: perspective_matrix.m41,
+                    m21: perspective_matrix.m12, m22: perspective_matrix.m22,
+                    m23: perspective_matrix.m32, m24: perspective_matrix.m42,
+                    m31: perspective_matrix.m13, m32: perspective_matrix.m23,
+                    m33: perspective_matrix.m33, m34: perspective_matrix.m43,
+                    m41: perspective_matrix.m14, m42: perspective_matrix.m24,
+                    m43: perspective_matrix.m34, m44: perspective_matrix.m44,
+                };
+
+                let mut tmp: [f32; 4] = [0.0; 4];
+                % for i in range(1, 5):
+                    tmp[${i - 1}] = (right_hand_side[0] * perspective_matrix.m1${i}) +
+                                                  (right_hand_side[1] * perspective_matrix.m2${i}) +
+                                                  (right_hand_side[2] * perspective_matrix.m3${i}) +
+                                                  (right_hand_side[3] * perspective_matrix.m4${i});
+                % endfor
+
+                Perspective(tmp[0], tmp[1], tmp[2], tmp[3])
+            } else {
+                Perspective(0.0, 0.0, 0.0, 1.0)
+            };
+
+            // Next take care of translation
+            let translate = Translate3D (
+                matrix.m31,
+                matrix.m32,
+                matrix.m33
+            );
+
+            // Now get scale and shear. 'row' is a 3 element array of 3 component vectors
+            let mut row: [[f32; 3]; 3] = [[0.0; 3]; 3];
+            % for i in range(1, 4):
+                row[${i - 1}][0] = matrix.m${i}1;
+                row[${i - 1}][1] = matrix.m${i}2;
+                row[${i - 1}][2] = matrix.m${i}3;
+            % endfor
+
+            // Compute X scale factor and normalize first row.
+            // TODO(canaltinova): I'm not sure what is "length(row[0])" means in spec, is it true?
+            let row0len = (row[0][0] * row[0][0] + row[0][1] * row[0][1] + row[0][2] * row[0][2]).sqrt();
+            let mut scale = Scale3D(row0len, 0.0, 0.0);
+            row[0] = [row[0][0] / row0len, row[0][1] / row0len, row[0][2] / row0len];
+
+            // Compute XY shear factor and make 2nd row orthogonal to 1st.
+            let mut skew = Skew(dot(row[0], row[1]), 0.0, 0.0);
+            row[1] = combine(row[1], row[0], 1.0, -skew.0);
+
+            // Now, compute Y scale and normalize 2nd row.
+            let row1len = (row[0][0] * row[0][0] + row[0][1] * row[0][1] + row[0][2] * row[0][2]).sqrt();
+            scale.1 = row1len;
+            row[0] = [row[0][0] / row1len, row[0][1] / row1len, row[0][2] / row1len];
+            skew.0 /= scale.1;
+
+            // Compute XZ and YZ shears, orthogonalize 3rd row
+            skew.1 = dot(row[0], row[2]);
```

Comment:
here and on the next 3 lines, shouldn't this all of those skew indexes be +1 (as they are 1 based instead of zero like the spec says) so: skew.2 and later -skew.2 and so on?